src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java

```diff
@@ -0,0 +1,159 @@
+package org.walkmod.javalang.compiler.types;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.*;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+
+import sun.misc.Resource;
+import sun.misc.URLClassPath;
+
+/**
+ * A modified URLClassPath that indexes the contents of classpath elements, for faster resource locating.
+ *
+ * The standard URLClassPath does a linear scan of the classpath for each resource, which becomes
+ * prohibitively expensive for classpaths with many elements.
+ */
+public class IndexedURLClassPath extends URLClassPath {
+
+    private URL[] urls;
+    private int lastIndexed = 0;
+    private static URL RT_JAR;
+
+    static {
+        // static block to resolve java.lang package classes
+        String[] bootPath = System.getProperties().get("sun.boot.class.path").toString()
+                .split(Character.toString(File.pathSeparatorChar));
+        for (String lib : bootPath) {
+            if (lib.endsWith("rt.jar")) {
+                File f = new File(lib);
+                try {
+                    RT_JAR = f.toURI().toURL();
+                } catch (MalformedURLException e) {
+                    throw new RuntimeException("The java.lang classes cannot be loaded", e.getCause());
+                }
+            }
+        }
+    }
+
+    public IndexedURLClassPath(URL[] urls) {
+        super(urls);
```

Comment:
Make copy of array?

Reply:
why do you suggest to copy this array?

Reply:
Deep mistrust ;-) The array is writable and we are creating an alias. I don't want to check the control flow to make sure all users behave and don't modify. Immutable data types or defensive copy make it possible to keep it private/unmodifyable. Parallel execution plans make that even more important.

Reply:
I have marked it as final in 5e6fb8984b1ca86bed46c0eaabda4bfb7f06648a
The copy is created by who provides the array ;)